file.go

```diff
@@ -0,0 +1,136 @@
+package main
+
+import (
+	"io/ioutil"
+	"os"
+	"fmt"
+	"net/http"
+	"io"
+	"path/filepath"
+	"bytes"
+	"archive/zip"
+	"strings"
+)
+
+// Download the zip file at the given URL to a temporary local directory.
+// Returns the asbolute path to the downloaded zip file.
+// IMPORTANT: You must call defer os.RemoveAll(dir) when done with this directory!
+func downloadGithubZipFile(repoOwner, repoName, gitTag, githubToken string) (string, *fetchError) {
+	// Create a temp directory
+	tempDir, err := ioutil.TempDir("", "")
+	if err != nil {
+		return "", newErr(err)
+	}
+
+	// Create an empty file to write to
+	file, err := os.Create(filepath.Join(tempDir, "repo.zip"))
+	if err != nil {
+		return "", newErr(err)
+	}
+	defer file.Close()
+
+	// Define the url
+	url := fmt.Sprintf("https://api.github.com/repos/%s/%s/zipball/%s", repoOwner, repoName, gitTag)
+
+	// Download the file, possibly using the GitHub oAuth Token
+	httpClient := &http.Client{}
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return "", newErr(err)
+	}
+
+	if githubToken != "" {
+		req.Header.Set("Authorization", fmt.Sprintf("token %s", githubToken))
+	}
+
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return "", newErr(err)
+	}
+	if resp.StatusCode != 200 {
+		// Convert the resp.Body to a string
+		buf := new(bytes.Buffer)
+		buf.ReadFrom(resp.Body)
+		respBody := buf.String()
+
+		return "", newError(500, fmt.Sprintf("Failed to download file at the url %s. Received HTTP Response %d. Body: %s", url, resp.StatusCode, respBody))
```

Comment:
I'd like to learn more about best practices with custom error structs in Golang. In this case, the `fetchError` struct I created has an `errorCode` property that doesn't *necessarily* correspond to HTTP response codes, but by convention happens to.  So, I'd rather be more explicit here about exactly what error code is happening here so that other calling programs can explicitly "catch" it.  I have a feeling this code's approach to error handling can be improved.